src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,182 @@
+      block(l, lPartitioner, context, i, j)
+    private def rightBlock(i: Int, j: Int, context: TaskContext): Matrix =
+      block(r, rPartitioner, context, i, j)
+
+    def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), Matrix)] = {
+      val row = split.index % rowBlocks
+      val col = split.index / rowBlocks
+      val rowsInThisBlock: Int = (if (row + 1 == rowBlocks && rowsRemainder != 0) rowsRemainder else rowsPerBlock)
+      val colsInThisBlock: Int = (if (col + 1 == colBlocks && colsRemainder != 0) colsRemainder else colsPerBlock)
+
+      val finalResult = DenseMatrix.zeros(rowsInThisBlock, colsInThisBlock)
+      var i = 0
+      while (i < nProducts) {
+        val leftMat = leftBlock(row, i, context)
+        val rightMat = rightBlock(i, col, context)
+        (leftMat, rightMat) match {
+          case (x: DenseMatrix, y: DenseMatrix) =>
+            fmaDenseMatricesWithExtension(x, y, finalResult)
+          case (x: DenseMatrix, y: SparseMatrix) =>
+            fmaDenseMatricesWithExtension(x, toDenseMatrix(y), finalResult)
+          case (x: SparseMatrix, y: DenseMatrix) =>
+            fmaDenseMatricesWithExtension(toDenseMatrix(x), y, finalResult)
+          case _ =>
+             throw new SparkException(s"I only support Dense * Dense, recieved: ${leftMat.getClass}, ${rightMat.getClass}.")
+        }
+
+        i += 1
+      }
+
+      Array(((row, col), finalResult)).iterator
+    }
+
+    /**
+      * Implemented by subclasses to return the set of partitions in this RDD. This method will only
+      * be called once, so it is safe to implement a time-consuming computation in it.
+      *
+      * The partitions in this array must satisfy the following property:
+      *   `rdd.partitions.zipWithIndex.forall { case (partition, index) => partition.index == index }`
+      */
+    protected def getPartitions: Array[Partition] =
+      partitions
+
+    /** Optionally overridden by subclasses to specify how they are partitioned. */
+    @transient override val partitioner: Option[Partitioner] =
+      Some(new GridPartitioner(rowBlocks, colBlocks, 1, 1))
+  }
+
+  case class IntPartition(index: Int) extends Partition
+
+  def multiply(l: BlockMatrix, r: BlockMatrix): BlockMatrix = {
+    require(l.numCols() == r.numRows(), "The number of columns of A and the number of rows " +
+      s"of B must be equal. A.numCols: ${l.numCols()}, B.numRows: ${r.numRows()}. If you " +
+      "think they should be equal, try setting the dimensions of A and B explicitly while " +
+      "initializing them.")
+
+    // FIXME: don't require same blocksize on each matrix
+    require(l.colsPerBlock == r.rowsPerBlock)
+    require(l.numColBlocks == r.numRowBlocks)
+
+    if (l.colsPerBlock == r.rowsPerBlock) {
+      val lPartitioner = l.blocks.partitioner.get.asInstanceOf[GridPartitioner]
+      val rPartitioner = r.blocks.partitioner.get.asInstanceOf[GridPartitioner]
```

Comment:
For testing purposes, I wrote some code to generate an rdd of ((Int, Int), random dense matrix). Calling multiply failed with a shitty error because I had never used a partitioner on my RDD. We should create a partitioner here if it doesn't have one.